google/cloud/aiplatform/jobs.py

```diff
@@ -15,10 +15,212 @@
+                project set in aiplatform.init will be used.
+            location: Optional[str] = None,
+                Optional location to retrieve BatchPredictionJob from. If not set,
+                location set in aiplatform.init will be used.
+            credentials: Optional[auth_credentials.Credentials] = None,
+                Custom credentials to use. If not set, credentials set in
+                aiplatform.init will be used.
+        """
+        valid_batch_prediction_job_name = utils.full_resource_name(
+            resource_name=batch_prediction_job_name,
+            resource_noun="batchPredictionJobs",
+            project=project,
+            location=location,
+        )
+
+        super().__init__(
+            valid_job_name=valid_batch_prediction_job_name,
+            project=project,
+            location=location,
+            credentials=credentials,
+        )
+
+    def iter_outputs(
+        self, bq_max_results: Optional[int] = 100
+    ) -> Union[Iterable[storage.Blob], Iterable[bigquery.table.RowIterator]]:
+        """Returns an Iterable object to traverse the output files, either a list
+        of GCS Blobs or a BigQuery QueryJob depending on the output config set
+        when the BatchPredictionJob was created.
+
+        Args:
+            bq_max_results: Optional[int] = 100
+                Limit on rows to retrieve from prediction table in BigQuery dataset.
+                Only used when retrieving predictions from a bigquery_destination_prefix.
+                Default is 100.
+
+        Returns:
+            Union[Iterable[storage.Blob], Iterable[bigquery.table.RowIterator]]:
+                Either a list of GCS Blob objects within the prediction output
+                directory or an iterable BigQuery RowIterator with predictions.
+
+        Raises:
+            RuntimeError:
+                If BatchPredictionJob is in a JobState other than SUCCEEDED,
+                since outputs cannot be retrieved until the Job has finished.
+            NotImplementedError:
+                If BatchPredictionJob succeeded and output_info does not have a
+                GCS or BQ output provided.
+        """
+
+        job_status = self.status()
+
+        if job_status != job_state.JobState.JOB_STATE_SUCCEEDED:
+            raise RuntimeError(
+                f"Cannot read outputs until BatchPredictionJob has succeeded, "
+                f"current status: {job_status}"
+            )
+
+        output_info = self._gca_resource.output_info
+
+        # GCS Destination, return Blobs
+        if output_info.gcs_output_directory:
+            storage_client = storage.Client()
+            blobs = storage_client.list_blobs(output_info.gcs_output_directory)
+            return blobs
+
+        # BigQuery Destination, return QueryJob
+        elif output_info.bigquery_output_dataset:
+            bq_client = bigquery.Client()
```

Comment:
I'm a bit concerned about this. Ideally you'd re-use the credentials from uCAIP client.

Also, there's a risk of leaking sockets when you create clients on-the-fly. Not as big a deal for REST clients, but definitely a concern for gRPC clients. https://github.com/googleapis/google-cloud-python/issues/9790 https://github.com/googleapis/google-cloud-python/issues/9457

Reply:
```suggestion
            bq_client = bigquery.Client(
                credentials=self.api_client._transport._credentials
            )
```

^ This change would build a BigQuery Client using the same credentials as uCAIP's `JobServiceClient`.

In regards to the leaking sockets, would [the solution referenced in that issue](https://github.com/googleapis/google-cloud-python/issues/9790#issuecomment-558263473) work? See below
```
# Close sockets opened by BQ Client
bq_client._http._auth_request.session.close()
bq_client._http.close()
```

Reply:
This change for credentials LGTM. (Storage should get similar treatment).

It's a little trickier in our case, because we want the client to live for the lifetime of the RowIterator.

Unless you want to convert to full list of rows / pandas dataframe before returning? In which case all the API requests would be made here and we could close the client when done (FWIW, the client does have a `close` function in BQ. https://googleapis.dev/python/bigquery/latest/generated/google.cloud.bigquery.client.Client.html#google.cloud.bigquery.client.Client.close)

Reply:
Made the credentials change on both BQ and Storage.

Re: closing connections - this is indeed tricky since the method is meant to return an iterator. However your comment made realize a larger issue of us instantiating a GAPIC client for every instance of a high-level SDK object. I'm capturing this in [b/174111905](http://b/174111905).

Will merge this blocking PR for now, thanks for calling this issue out!